### DISCUSSION TITLE
基于 NebulaGraph 图数据库的欺诈检测方法与代码示例 - siwei.io

### DISCUSSION THREAD

**Discussion by giscus** [siwei-io/siwei-io.github.io#8] [Comments]
# 基于 NebulaGraph 图数据库的欺诈检测方法与代码示例 - siwei.io

Siwei(思为) builds things and believes in Open Source.

https://www.siwei.io/fraud-detection-with-nebulagraph/

**Comment by wey-gu** [comment]
Update:

GNN 部分的示例项目的代码在：[github.com/wey-gu/NebulaGraph-Fraud-Detection-GNN](https://github.com/wey-gu/NebulaGraph-Fraud-Detection-GNN) ，如有问题欢迎留言、ISSUE 纠错。

**Comment by wey-gu** [comment]
Update:

## 训练和线上推理系统的 notebook

| Notebook                                                     | Arch. and Flow                                               |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| [Train_GraphSAGE.ipynb](https://github.com/wey-gu/NebulaGraph-Fraud-Detection-GNN/tree/main/notebooks/Train_GraphSAGE.ipynb) | <img style="float: middle;" src="https://user-images.githubusercontent.com/1651790/182063699-f2f982e0-1dc4-437a-b509-f6247feb475b.svg"> |
| [Inference_API.ipynb](https://github.com/wey-gu/NebulaGraph-Fraud-Detection-GNN/tree/main/notebooks/Inference_API.ipynb) | ![GraphSAGE_FraudDetection_Inference](https://user-images.githubusercontent.com/1651790/182292372-2bef1e38-db4e-4949-8f66-bff361ee93d9.svg) |

## 线上系统全栈例子代码

![](https://user-images.githubusercontent.com/1651790/182621458-702fa2be-bdd9-4a67-8288-51503a3a97da.svg)

Demo Video 👉🏻:

https://user-images.githubusercontent.com/1651790/182651965-d489a218-36a6-40c9-9fab-ba288e8d959a.mov

Code 👉🏻:

https://github.com/wey-gu/NebulaGraph-Fraud-Detection-GNN/tree/main/src

**Comment by kinggggg** [comment]
下面的查询中，`<-[e:with_phone_num]-(:applicant)`部分是多余的吧？
> 

```cypher
## 针对一笔交易申请关联查询
MATCH (n) WHERE id(n) == "200000010265"
OPTIONAL MATCH p_shared_d=(n)-[:used_device]->(d)<-[:used_device]-(:applicant)-[:with_phone_num]->(pn:phone_num)<-[e:with_phone_num]-(:applicant)
RETURN p_shared_d
```

**Reply by wey-gu** [reply]
👍🏻，的确，不是必须的，不过这个query表达的是：和这个人共享过设备（wifi/手机）的申请人共享过手机号的申请人。
这个例子的的确确不是一个最小、最简洁的哈。

**Comment by haojiangt** [comment]
大佬您好，想请教您几个问题。您最后说的这个：“定期计算图谱中的图特征，和其他特征一起用传统机器学习方法离线预测风险”是使用图算法（如图神经网络）嵌入一些新特征，然后与原特征进行拼接后再输入到传统机器学习模型（如XGBoost）中进行检测是嘛，我目前想往这方面研究一下，从而完成我的硕士毕业论文。
想咨询您一下：1.您有这种类似的相关数据集推荐嘛？
2.这样效果会好吗？或者有没有推荐的方法，或者所谓的“创新呢”
3.您有推荐的相关论文嘛，因为我很少找到像我这种想法的文章。
谢谢您。

**Reply by wey-gu** [reply]
您好 :)

> “定期计算图谱中的图特征，和其他特征一起用传统机器学习方法离线预测风险”是使用图算法（如图神经网络）嵌入一些新特征，然后与原特征进行拼接后再输入到传统机器学习模型（如XGBoost）中进行检测是嘛，我目前想往这方面研究一下，从而完成我的硕士毕业论文。

这个方法好像已经是非常非常常见的方法了哈，前些年就有了，可以去 google 学术搜一下哈，最近，比如这个论文 https://eprint.iacr.org/2018/917.pdf 是在讨论怎么在没法获得全图的情况下去算 pagerank 你可以看下挺有意思的呢。
加油，💪🏻，现在经济下行，用技术而非人工、专家的 fraud detection 越来越被人重视了，是一个好方向。

实操的话参考这个，好像是一个学生上课的作业：https://github.com/aravind-sundaresan/Graph-ML-Fraud-Detection 

> 1.您有这种类似的相关数据集推荐嘛？

https://www.kaggle.com/ntnu-testimon/banksim1 这个用的挺多的哈

> 2.这样效果会好吗？或者有没有推荐的方法，或者所谓的“创新呢”

看比较是会好的，这种方法挺多的，比如简单的逻辑回归，带上 pagerank 一个 feature 就会比不带上的效果好一些

> 3. 您有推荐的相关论文嘛，因为我很少找到像我这种想法的文章。

看看这个？ https://www.researchgate.net/profile/Rafael-Van-Belle/publication/338364320_Representation_Learning_in_Graphs_for_Credit_Card_Fraud_Detection/links/5e8320b6a6fdcca789e230f8/Representation-Learning-in-Graphs-for-Credit-Card-Fraud-Detection.pdf

**Reply by haojiangt** [reply]
非常非常感谢您的回答，我再从谷歌学术查找一下，学习一下您推荐的这些文章与作业，如果可以的话能否留下您的联系方式，或者其它类似公众号的东西，以便于我更方便的向您学习，再次感谢您的回答。

**Reply by wey-gu** [reply]
好滴好滴👍

这里可以找到我的联系方式哈👉

[https://siwei.io/about/#联系我](https://www.siwei.io/about/#%e8%81%94%e7%b3%bb%e6%88%91)

**Reply by haojiangt** [reply]
因为我不擅长用推特和其它联系方式，而您的微信id似乎已经更改使我无法查找到您，如果是因为个人原因不方便添加微信，那么我将继续跟随您的github和siwei.io进行学习，再次感谢您的文章和回答，它使我受益匪浅。

**Reply by wey-gu** [reply]
> 因为我不擅长用推特和其它联系方式，而您的微信id似乎已经更改使我无法查找到您，如果是因为个人原因不方便添加微信，那么我将继续跟随您的github和siwei.io进行学习，再次感谢您的文章和回答，它使我受益匪浅。

哈哈，你可能没看清楚，我给了微信在里边，你在 shell 里执行这个就能获得我的微信 id，而不是 `c2l2dmVpCg==`

> 
`echo c2l2dmVpCg== | base64 -d`

或者你不方便，可以给我发邮件告诉你的 微信 id 我加你。

**Reply by haojiangt** [reply]
哈哈哈，非常抱歉，我比较愚笨，没有发现，我已经成功获取并添加您的微信，非常感谢您。

**Comment by wey-gu** [comment]
update:

因为后边数据集生成进行了重构，在新的 sample 数据集里，这个 query 是一个例子

```cypher
MATCH (n) WHERE id(n) == "p_8591"
OPTIONAL MATCH p_shared_d=(n)-[:`used_device`]->(d)<-[:`used_device`]-(:`applicant`)-[:`with_phone_num`]->(pn:phone_num)<-[e:`with_phone_num`]-(:`applicant`)
RETURN p_shared_d LIMIT 500
```

![](https://user-images.githubusercontent.com/1651790/239861937-3e2adcaf-aa8b-4519-ab2a-ba35caf9ccb4.png)

**Comment by xinzaifeixiang1992** [comment]
liunx单机nebula 3.5.0，与nebula-studio 3.7.0可以搭配使用；但是跟nebula importer 4.0.0就不兼容了。直接启动importer之后，指定--config 路径，不能进行数据导入，报错为 unsupported client version。
nebula 3.5.0通过importer工具进行导入的话，适配的版本是哪个版本呀?

**Reply by wey-gu** [reply]
NebulaGraph 3.5.0 的 importer 可以适配 NebulaGraph 3.x.x 的所有数据库版本哈

**Comment by xinzaifeixiang1992** [comment]
我将古老师项目中的示例数据拉到liunx单机nebula 3.5.0机器上了，将数据导入图数据库中时，修改完nebula_graph_importer.yaml文件中的路径和对应的graphd的ip后，直接命令行启动nebula-importer，后面配置--config **.yaml，可以按照这种方式导入数据吗

**Reply by wey-gu** [reply]
没错，可以的哈
